File: engine/src/router.cpp
#include "antsql/router.hpp"

#include <algorithm>
#include <cmath>
#include <functional>
#include <limits>

namespace antsql {

std::size_t RouteKeyHash::operator()(const RouteKey& key) const noexcept {
  return std::hash<std::string>{}(key.table) ^ (std::hash<std::uint32_t>{}(key.partition) << 1U);
}

Router::Router(RouterConfig config) : config_(config), rng_(config.random_seed) {}

double Router::Pheromone(const RouteKey& key, const std::string& neighbor) const {
  const auto route = pheromone_.find(key);
  if (route == pheromone_.end()) return config_.initial_pheromone;
  const auto score = route->second.find(neighbor);
  return score == route->second.end() ? config_.initial_pheromone : score->second;
}

double Router::Score(const RouteKey& key, const Neighbor& candidate) const {
  return Pheromone(key, candidate.id) / std::max(candidate.link_latency_ms, 0.001);
}

std::optional<RouteDecision> Router::Choose(
    const RouteKey& key, const std::vector<Neighbor>& candidates,
    const std::unordered_set<std::string>& visited) {
  std::vector<Neighbor> eligible;
  std::copy_if(candidates.begin(), candidates.end(), std::back_inserter(eligible),
               [&visited](const Neighbor& neighbor) { return !visited.contains(neighbor.id); });
  if (eligible.empty()) return std::nullopt;

  std::bernoulli_distribution explore(config_.exploration_probability);
  if (explore(rng_)) {
    std::uniform_int_distribution<std::size_t> pick(0, eligible.size() - 1);
    const auto& chosen = eligible[pick(rng_)];
    return RouteDecision{chosen.id, Score(key, chosen), true};
  }

  const auto best = std::max_element(eligible.begin(), eligible.end(),
      [this, &key](const Neighbor& lhs, const Neighbor& rhs) { return Score(key, lhs) < Score(key, rhs); });
  return RouteDecision{best->id, Score(key, *best), false};
}

void Router::ObserveSuccess(const RouteKey& key, const std::vector<std::string>& path,
                            double total_cost_ms) {
  if (path.size() < 2) return;
  const auto quality = config_.reference_cost_ms / std::max(total_cost_ms, 0.001);
  const auto reinforcement = config_.base_reinforcement * std::clamp(
      quality, config_.minimum_reinforcement_fraction, 1.0);
  auto& scores = pheromone_[key];
  for (std::size_t index = 0; index + 1 < path.size(); ++index) {
    scores[path[index + 1]] += reinforcement;
  }
}

void Router::ObserveFailure(const RouteKey& key, const std::vector<std::string>& path) {
  if (path.size() < 2) return;
  auto& scores = pheromone_[key];
  const auto& failed_next_hop = path.back();
  scores[failed_next_hop] = std::max(0.0, Pheromone(key, failed_next_hop) * config_.failure_penalty);
}

void Router::Evaporate() {
  for (auto& [_, scores] : pheromone_) {
    for (auto& [__, amount] : scores) amount *= (1.0 - config_.evaporation_rate);
  }
}

}  // namespace antsql

File: service/server/router.js
'use strict';

// A JS port of engine/include/antsql/router.hpp's pheromone-routing
// algorithm (same fields, same formulas), applied here to replica
// selection instead of gateway-to-gateway hops. Kept as a faithful port
// rather than a from-scratch reimplementation so the product's behavior
// stays traceable to the research engine it's built on.
class Router {
  constructor(config = {}) {
    this.config = {
      evaporationRate: 0.02,
      baseReinforcement: 2.5,
      referenceCostMs: 8.0,
      minimumReinforcementFraction: 0.02,
      failurePenalty: 0.25,
      explorationProbability: 0.06,
      initialPheromone: 0.5,
      ...config,
    };
    // routeKey (string) -> Map(candidateId -> pheromone score)
    this.pheromone = new Map();
  }

  _routeMap(key) {
    let m = this.pheromone.get(key);
    if (!m) {
      m = new Map();
      this.pheromone.set(key, m);
    }
    return m;
  }

  pheromoneFor(key, candidateId) {
    const m = this.pheromone.get(key);
    if (!m || !m.has(candidateId)) return this.config.initialPheromone;
    return m.get(candidateId);
  }

  _score(key, candidate) {
    return this.pheromoneFor(key, candidate.id) / Math.max(candidate.linkLatencyMs, 0.001);
  }

  // candidates: [{id, linkLatencyMs}]; visited: Set<id> to exclude.
  choose(key, candidates, visited = new Set()) {
    const eligible = candidates.filter((c) => !visited.has(c.id));
    if (eligible.length === 0) return null;

    if (Math.random() < this.config.explorationProbability) {
      const chosen = eligible[Math.floor(Math.random() * eligible.length)];
      return { nextHop: chosen.id, score: this._score(key, chosen), explored: true };
    }

    let best = eligible[0];
    let bestScore = this._score(key, best);
    for (const candidate of eligible.slice(1)) {
      const score = this._score(key, candidate);
      if (score > bestScore) {
        best = candidate;
        bestScore = score;
      }
    }
    return { nextHop: best.id, score: bestScore, explored: false };
  }

  observeSuccess(key, candidateId, totalCostMs) {
    const quality = this.config.referenceCostMs / Math.max(totalCostMs, 0.001);
    const reinforcement =
      this.config.baseReinforcement *
      Math.min(1.0, Math.max(this.config.minimumReinforcementFraction, quality));
    const m = this._routeMap(key);
    m.set(candidateId, this.pheromoneFor(key, candidateId) + reinforcement);
  }

  observeFailure(key, candidateId) {
    const m = this._routeMap(key);
    m.set(candidateId, Math.max(0, this.pheromoneFor(key, candidateId) * this.config.failurePenalty));
  }

  evaporate() {
    for (const m of this.pheromone.values()) {
      for (const [candidateId, score] of m) m.set(candidateId, score * (1 - this.config.evaporationRate));
    }
  }
}

module.exports = { Router };
